jose/src/jws/crypto/mod.rs

```diff
@@ -0,0 +1,152 @@
+// SPDX-FileCopyrightText: 2022 Profian Inc. <opensource@profian.com>
+// SPDX-License-Identifier: Apache-2.0
+
+//! JWS Cryptographic Implementation
+
+pub mod rcrypto;
+
+mod core;
+mod sig;
+
+pub use sig::State;
+
+use alloc::{vec, vec::Vec};
+
+use crate::b64::Update;
+use crate::jws::{Protected, Signature, Unprotected};
+
+use super::{Flattened, General, Jws};
+
+/// Signature creation state
+pub trait Signer: Update {
+    #[allow(missing_docs)]
+    type FinishError: From<Self::Error>;
+
+    /// Finish processing payload and create the signature.
+    fn finish(self) -> Result<Signature, Self::FinishError>;
+}
+
+/// A signature creation key
+pub trait SigningKey<'a> {
+    #[allow(missing_docs)]
+    type StartError: From<<Self::Signer as Update>::Error>;
+
+    /// The state object used during signing.
+    type Signer: Signer;
+
+    /// Begin the signature creation process.
+    fn sign(
+        &'a self,
+        prot: Option<Protected>,
+        head: Option<Unprotected>,
+    ) -> Result<Self::Signer, Self::StartError>;
+}
+
+/// Signature verification state
+pub trait Verifier<'a>: Update {
```

Comment:
If you got a `Verifier`, why not name the `SigningKey` trait `Signer`? This makes more sense cuz it can also adapt to HSM etc.

Reply:
`Verifier` represents the state of an in-process streamed verification. `VerifyingKey` represents the object that can start the verification process.

`Signer` represents the state of an in-process streamed signature creation. `SigningKey` represents the object that can start the signature creation process.